File: src/dataset/python/markdown_exercism-fruit-in/openAI/ChatGPT4_markdown_v3.py
import re


def parse(markdown):
    lines = []
    start_list = False

    for line in markdown.splitlines():
        parsed_line = parse_line(line)
        if not start_list and parsed_line.startswith('<li>'):
            lines.append('<ul>')
            start_list = True
        if start_list and not parsed_line.startswith('<li>'):
            lines.append('</ul>')
            start_list = False
        lines.append(parsed_line)

    if start_list:
        lines.append('</ul>')

    return ''.join(lines)


def parse_line(markdown):
    markdown = emphasis(markdown)

    if is_header(markdown):
        i = markdown.find(' ')
        return f'<h{i}>{markdown[i + 1:]}</h{i}>'
    elif is_unordered_list(markdown):
        return f'<li>{markdown[2:]}</li>'
    else:
        return f'<p>{markdown}</p>'


def emphasis(markdown):
    markdown = re.sub(r'__(.*?)__', r'<strong>\1</strong>', markdown)
    markdown = re.sub(r'_(.*?)_', r'<em>\1</em>', markdown)
    return markdown


def is_header(markdown):
    return markdown.startswith('#') and ' ' in markdown


def is_unordered_list(markdown):
    return markdown.startswith('* ')

File: src/dataset/Java/markdown_exercism-shyvum/openAI/ChatGPT4_Markdown_v3.java
class Markdown {

    String parse(String markdown) {
        String[] lines = markdown.split("\n");
        StringBuilder result = new StringBuilder();
        boolean activeList = false;

        for (String line : lines) {
            String parsedLine = parseHeader(line);

            if (parsedLine == null) {
                parsedLine = parseListItem(line);
            }

            if (parsedLine == null) {
                parsedLine = parseParagraph(line);
            }

            if (parsedLine.startsWith("<li>") && !activeList) {
                activeList = true;
                result.append("<ul>");
            } else if (!parsedLine.startsWith("<li>") && activeList) {
                activeList = false;
                result.append("</ul>");
            }

            result.append(parsedLine);
        }

        if (activeList) {
            result.append("</ul>");
        }

        return result.toString();
    }

    private String parseHeader(String markdown) {
        int count = 0;
        while (count < markdown.length() && markdown.charAt(count) == '#') {
            count++;
        }

        if (count == 0) {
            return null;
        }

        return "<h" + count + ">" + markdown.substring(count + 1).trim() + "</h" + count + ">";
    }

    private String parseListItem(String markdown) {
        if (markdown.startsWith("* ")) {
            return "<li>" + parseSomeSymbols(markdown.substring(2).trim()) + "</li>";
        }
        return null;
    }

    private String parseParagraph(String markdown) {
        return "<p>" + parseSomeSymbols(markdown.trim()) + "</p>";
    }

    private String parseSomeSymbols(String markdown) {
        return markdown.replaceAll("__(.+?)__", "<strong>$1</strong>")
                       .replaceAll("_(.+?)_", "<em>$1</em>");
    }
}